CAN: STM32H7: can2 interface is not usable

**Describe the bug**
The can2 interface as defined in the STM32H7 device tree is not able to send or receive CAN frames.

I use a custom board, based on the STM32H743VI, with two CAN interfaces and have issues to make them work simultaneously.
I wrote a test program that sets up threads and messages queues, in the same way for each CAN interface (can1 and can2) so that any CAN frame received on one interface is sent back on the same interface while incrementing the first data byte of the payload.

The frames are in the extended format, and the receiving is handled with a filter that accepts all data frames in extended format, and uses a message queue.

Whether I connect the two interfaces together or use a setup with two boards, I always get the same result: can1 is able to send and receive CAN frames while can2 is not.

However, can2 works if I disable the can1 interface and alter the can2 interface definition in the STM32H7 device tree by replacing these two lines:
```
reg = <0x4000a400 0x400>, <0x4000ac00 0x6a0>;
bosch,mram-cfg = <0x350 28 8 3 3 0 3 3>;
```
with these two lines:
```
reg = <0x4000a400 0x400>, <0x4000ac00 0x350>;
bosch,mram-cfg = <0x0 28 8 3 3 0 3 3>;
```

This makes me think there could be an issue in the CAN driver related to the message RAM layout.

**To Reproduce**
- Set the can2 interface to okay in the STM32H7 device tree
- Build an app that receives all extended data frames on can2 and sends them back
- Connect the can2 interface to another working CAN interface that sends extended data frames, while making sure the bus speeds match on the two interfaces
- Observe that no CAN frame is received on CAN2

**Expected behavior**
can2 is able to send and receive CAN frames, while simultaneously using the can1 interface.

**Impact**
can2 is not usable on my board.

I've tried adjusting the addresses and offsets for can2 without success:

If I use ```reg = <0x4000a400 0x400>, <0x4000af50 0x350>;``` and  ```bosch,mram-cfg = <0x0 28 8 3 3 0 3 3>;``` the interface still does not work.
If I use ```reg = <0x4000a400 0x400>, <0x4000af50 0x350>;``` and  ```bosch,mram-cfg = <0x350 28 8 3 3 0 3 3>;```, I get the following build error: 
```
zephyr/drivers/can/can_stm32h7.c:197:9: error: static assertion failed: "Insufficient Message RAM size to hold elements"
  197 |         BUILD_ASSERT(CAN_MCAN_DT_INST_MRAM_ELEMENTS_SIZE(n) <=              \
```

> I've tried adjusting the addresses and offsets for can2 without success:
> 
> If I use `reg = <0x4000a400 0x400>, <0x4000af50 0x350>;` and `bosch,mram-cfg = <0x0 28 8 3 3 0 3 3>;` the interface still does not work. If I use `reg = <0x4000a400 0x400>, <0x4000af50 0x350>;` and `bosch,mram-cfg = <0x350 28 8 3 3 0 3 3>;`, I get the following build error:
> 
> ```
> zephyr/drivers/can/can_stm32h7.c:197:9: error: static assertion failed: "Insufficient Message RAM size to hold elements"
>   197 |         BUILD_ASSERT(CAN_MCAN_DT_INST_MRAM_ELEMENTS_SIZE(n) <=              \
> ```

Right. With the configuration above, you have reduced the Bosch M_CAN message RAM size to zero bytes, hence the number of elements specified will not fit.

--- a/drivers/can/can_stm32h7.c
+++ b/drivers/can/can_stm32h7.c
@@ -21,6 +21,7 @@ LOG_MODULE_REGISTER(can_stm32h7, CONFIG_CAN_LOG_LEVEL);
 
 struct can_stm32h7_config {
 	mm_reg_t base;
+	mem_addr_t mrba;
 	mem_addr_t mram;
 	void (*config_irq)(void);
 	const struct pinctrl_dev_config *pcfg;
@@ -132,7 +133,7 @@ static int can_stm32h7_init(const struct device *dev)
 		return ret;
 	}
 
-	ret = can_mcan_configure_mram(dev, stm32h7_cfg->mram, stm32h7_cfg->mram);
+	ret = can_mcan_configure_mram(dev, stm32h7_cfg->mrba, stm32h7_cfg->mram);
 	if (ret != 0) {
 		return ret;
 	}
@@ -205,6 +206,7 @@ static const struct can_mcan_ops can_stm32h7_ops = {
 									    \
 	static const struct can_stm32h7_config can_stm32h7_cfg_##n = {	    \
 		.base = CAN_MCAN_DT_INST_MCAN_ADDR(n),			    \
+		.mrba = CAN_MCAN_DT_INST_MRBA(n),			    \
 		.mram = CAN_MCAN_DT_INST_MRAM_ADDR(n),			    \
 		.config_irq = stm32h7_mcan_irq_config_##n,		    \
 		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(n),		    \

--- a/include/zephyr/drivers/can/can_mcan.h
+++ b/include/zephyr/drivers/can/can_mcan.h
@@ -573,13 +573,25 @@
  * @brief Get the Bosch M_CAN Message RAM base address
  *
  * For devicetree nodes with dedicated Message RAM area defined via devicetree, this macro returns
- * the base address of the Message RAM, taking in the Message RAM offset into account.
+ * the base address of the Message RAM.
  *
  * @param node_id node identifier
- * @return the Bosch M_CAN Message RAM base address
+ * @return the Bosch M_CAN Message RAM base address (MRBA)
+ */
+#define CAN_MCAN_DT_MRBA(node_id)                                                                  \
+	(mem_addr_t)DT_REG_ADDR_BY_NAME(node_id, message_ram)
+
+/**
+ * @brief Get the Bosch M_CAN Message RAM address
+ *
+ * For devicetree nodes with dedicated Message RAM area defined via devicetree, this macro returns
+ * the address of the Message RAM, taking in the Message RAM offset into account.
+ *
+ * @param node_id node identifier
+ * @return the Bosch M_CAN Message RAM address
  */
 #define CAN_MCAN_DT_MRAM_ADDR(node_id)                                                             \
-	(mem_addr_t)(DT_REG_ADDR_BY_NAME(node_id, message_ram) + CAN_MCAN_DT_MRAM_OFFSET(node_id))
+	(mem_addr_t)(CAN_MCAN_DT_MRBA(node_id) + CAN_MCAN_DT_MRAM_OFFSET(node_id))
 
 /**
  * @brief Get the Bosch M_CAN Message RAM size
@@ -783,9 +795,17 @@
 #define CAN_MCAN_DT_INST_MCAN_ADDR(inst) CAN_MCAN_DT_MCAN_ADDR(DT_DRV_INST(inst))
 
 /**
+ * @brief Equivalent to CAN_MCAN_DT_MRBA(DT_DRV_INST(inst))
+ * @param inst DT_DRV_COMPAT instance number
+ * @return the Bosch M_CAN Message RAM Base Address (MRBA)
+ * @see CAN_MCAN_DT_MRBA()
+ */
+#define CAN_MCAN_DT_INST_MRBA(inst) CAN_MCAN_DT_MRBA(DT_DRV_INST(inst))
+
+/**
  * @brief Equivalent to CAN_MCAN_DT_MRAM_ADDR(DT_DRV_INST(inst))
  * @param inst DT_DRV_COMPAT instance number
- * @return the Bosch M_CAN Message RAM base address
+ * @return the Bosch M_CAN Message RAM address
  * @see CAN_MCAN_DT_MRAM_ADDR()
  */
 #define CAN_MCAN_DT_INST_MRAM_ADDR(inst) CAN_MCAN_DT_MRAM_ADDR(DT_DRV_INST(inst))

